### PR TITLE
FPS tutorial remaster

### PR DESCRIPTION
**Pull request by TwistedTwigleg**
______

I'm remastering the FPS tutorial so it has the following:

* Zip files for each and every part of the tutorial
* A consistent code/node style.
* Two more parts (the code for the final part is located [here](https://github.com/TwistedTwigleg/Godot_FPS_Tutorial) )
* Rewritten text that is (hopefully) easier to understand

Progress:

- [x] Rewrite/Update part 1
- [x] Rewrite/Update part 2
- [x] Rewrite/Update part 3
- [x] Remove the old part 4 (since the modular code is now used in parts 1-3)
- [x] Add part 4
- [x] Add part 5
- [x] Add part 6
- [x] Update/Add zip files for part 1
- [x] Update/Add zip files for part 2
- [x] Update/Add zip files for part 3
- [x] Update/Add zip files for part 4
- [x] Update/Add zip files for part 5
- [x] Update/Add zip files for part 6

*(I have not had a chance to go and edit just yet, so there's likely a fair amount of mistakes. I will give it a editing pass soon!)*

Edit: I'll squish some/all of the commits once everything is done

Edit 2: Split part 5 into two parts because it would have been really long otherwise.
____

Any/All feedback is welcome!

### CODE REVIEW

**Comment by TwistedTwigleg** [comment]
Thanks @Noshyaar!
I (hopefully) fixed the mistakes.

File: tutorials/3d/fps_tutorial/part_one.rst [hunk 1]
```diff
@@ -8,15 +8,37 @@ Tutorial introduction
...
 
+.. error:: TODO: redo this image!
+
 This tutorial series will show you how to make a single player FPS game.
 
-Throughout the course of these tutorials, we will cover how:
+Throughout the course of this tutorial series, we will cover how:
 
-- To make a first person character, with sprinting and a flash light.
+- To make a first person character that can:
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_one.rst]: Putting some of the bullets below right after this line feels weird
```
To make a first person character that can: Move around the environment      (okay)
To make a first person character that can: That can jump into the air ...   (redundant)
To make a first person character that can: That has a flash light we ...    (hmm)
and so on
```

I'm not sure if I explain clearly or even correctly, but maybe you could do something with it. :D

File: tutorials/3d/fps_tutorial/part_one.rst [hunk 2]
```diff
@@ -31,21 +53,21 @@ Throughout the course of these tutorials, we will cover how:
 
 You can find the start assets for this parts 1 through 3 here: :download:`Godot_FPS_Starter.zip <files/Godot_FPS_Starter.zip>`
 
-.. warning:: A video version of this tutorial series is coming soon!
+.. error:: TODO: update this zip file!
 
 The provided starter assets contain a animated 3D model, a bunch of 3D models for making levels,
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_one.rst]: > a animated

File: tutorials/3d/fps_tutorial/part_one.rst [hunk 3]
```diff
@@ -71,28 +94,18 @@ First, go open the project settings and go to the "Input Map" tab. You'll find s
...
-.. tip:: The reason we get those red lines is because we are using lowpoly models
-        with low resolution textures. MSAA tries to reduce jagged edges between models and
-        because we are using lowpoly and low resolution textures in this project,
-        we need to turn it off to avoid texture bleeding.
-
-        A bonus with turning off MSAA is we get a more 'retro' looking result.
-
 _________
 
 Lets take a second to see what we have in the starter assets.
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_one.rst]: should be `Let's`

File: tutorials/3d/fps_tutorial/part_one.rst [hunk 4]
```diff
@@ -71,28 +94,18 @@ First, go open the project settings and go to the "Input Map" tab. You'll find s
...
 
-Included in the starter assets are five scenes: ``BulletScene.tscn``, ``Player.tscn``,
-``SimpleAudioPlayer.tscn``, ``TestingArea.tscn``, and ``TestLevel.tscn``.
+Included in the starter assets are several scenes. For example, in `res://` we have 14 scenes, most of which we will be visiting as
+we go through this tutorial series.
 
-We will visit all of these scenes later, but for now open up ``Player.tscn``.
+For now lets open up ``Player.tscn``.
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_one.rst]: should be `let's`

Comment (Noshyaar) [reply tutorials/3d/fps_tutorial/part_one.rst]: There's too much to review. I won't be able to look at all of it at once. I'll revisit this some time and I'll let you know if I notice. Thank you very much for contributing.

Comment (TwistedTwigleg) [reply tutorials/3d/fps_tutorial/part_one.rst]: Thanks for reviewing! I'll fix these soon :)

**Comment by TwistedTwigleg** [comment]
Phew, all six parts now have first drafts. Now I just need to update the images, add the zip files, and edit the written material. It's getting there!

I definitely think I can have everything ready by the end of April, assuming I can keep going at this rate.

**Comment by TwistedTwigleg** [comment]
Alright, everything is done 🎉 😄 

It probably could use some editing passes to fix spelling mistakes, but I'm happy enough with where it is now to remove the WIP tag.

(Guess I should tag @TwistedTwigleg for review... No doubt he'll say it's all terrible 😉 )

File: tutorials/3d/fps_tutorial/part_four.rst [hunk 5]
```diff
@@ -6,1295 +6,728 @@ Part 4
...
+     
+     
+Let's go over what's happening:
 
-Lets look at ``fire_weapon``.
+First we check the mouse mode. If the mouse mode is not ``MOUSE_MODE_CAPTURED``, we want to return, which will skip the code below.
 
-First we make a clone of the bullet scene and add it as a child of the scene root.
-Next we set its global transform to ``self.global_transform``.
+Next we define a new :ref:`Vector2 <class_Vector2>` called ``joypad_vec``. This will hold the right joystick position. Based on the OS, we set it's values so
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: >we set it's values so

File: tutorials/3d/fps_tutorial/part_four.rst [hunk 6]
```diff
@@ -6,1295 +6,728 @@ Part 4
...
+Open up ``Player.gd`` and add the following global variables:
 
-If we do not have enough ammo in spares, we instead add all of the ammo left in spares and then set our spare ammo to zero.
+::
+    
+    var mouse_scroll_value = 0
+    const MOUSE_SENSITIVITY_SCROLL_WHEEL = 0.08
 
-We then play the reloading animation and play a sound. We return ``true`` to signal we have successfully reloaded.
+Lets go over what each of these new variables will be doing:
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: >Lets

File: tutorials/3d/fps_tutorial/part_four.rst [hunk 7]
```diff
@@ -6,1295 +6,728 @@ Part 4
...
 
-If we cannot reload because ``reload_weapon`` is ``false``, we return ``false`` to signal we did not successfully reload.
+* ``mouse_scroll_value``: The value of the mouse scroll wheel.
+* ``MOUSE_SENSITIVITY_SCROLL_WHEEL``: How much a single scroll action increases mouse_scroll_value
 
 ______
 
-For ``equip_weapon`` we first check if the player is in the pistol's idle state.
+Now lets add the following to ``_input``:
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: >lets

File: tutorials/3d/fps_tutorial/part_four.rst [hunk 8]
```diff
@@ -6,1295 +6,728 @@ Part 4
...
-.. tip:: When we are using ``for X in Y`` where ``Y`` is a dictionary, ``X`` is assigned to the each **key** in the dictionary, not the value. To get the value, we
-          have to retrieve it using ``Y[X]``.
+    func trigger_body_entered(body):
+        if body.has_method("add_health"):
+            body.add_health(HEALTH_AMOUNTS[kit_size])
+            respawn_timer = RESPAWN_TIME
+            kit_size_change_values(kit_size, false)
 
-If the weapon node is not ``null``, we set it's ``player_node`` variable to ``self``, and we make the point look at the gun aim position.
+Let's go over what this script is doing, starting with it's global variables:
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: >starting with it's global variables

File: tutorials/3d/fps_tutorial/part_four.rst [hunk 9]
```diff
@@ -6,1295 +6,728 @@ Part 4
...
+We get the collision shape for the node corresponding to ``size`` and disable it based on the ``enabled`` passed in argument/variable.
 
-Next we check get the current animation state from our animation manager, and we set ``is_reloading`` to ``false``.
-The reason we need ``is_reloading`` is because we need to go through each weapon and make sure we are not in it's reloading state already,
-because we do not want to allow the player to (potentially) reload if they are already in a reloading animation.
+.. note:: Why are we using ``!enable`` instead of ``enable``? This is so when we say we want to enable the node, we can pass in ``true``, but since
+          :ref:`CollisionShape <class_CollisionShape>` uses disabled instead of enabled, we need to flip it. By flipping it, we can enable the collision shape
+          and make the mesh visible when ``true`` is passed in.
+  
+We then get the correct :ref:`Spatial <class_Spatial>` node holding the mesh and set it's visibility to ``enable``.
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: >and set it's visibility

File: tutorials/3d/fps_tutorial/part_four.rst [hunk 10]
```diff
@@ -6,1295 +6,728 @@ Part 4
...
+While adding health is good and all, we can't really reap the rewards from it since nothing can (currently) damage us.
+Let's add some ammo pick ups next!
 
-Adding ``process_reloading``
-____________________________
+Open up ``Ammo_Pickup.tscn``. Notice how it's structured exactly the same as ``Health_Pickup.tscn``, just with the meshes and trigger collision shapes changed slightly to adjust
+for the difference in mesh sizes.
 
-Let's finish up our new modular weapon system and add ``process_reloading``. Make a new function called ``process_reloading`` and add the following:
+Select ``Ammo_Pickup`` and add a new script called ``A
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: `$"../NodeName"` returns sibling node.

Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_four.rst]: $"../NodeName" returns sibling node.

Comment (Noshyaar) [reply tutorials/3d/fps_tutorial/part_four.rst]: It's at the line 659. Github messed it up. :P

**Comment by TwistedTwigleg** [comment]
Thanks @Noshyaar! I fixed all of the issues you brought up.

I did a quick editing pass on part 3, so hopefully it's now more or less up to snuff.

Also, I decided to include a note as to why `get_node().get_parent()` is used instead of `$"../NodeName"`, since I didn't feel the change was big/important enough to go through and update all of the zip files. I just explained I didn't know about it at the time and showed how you can use `$` instead.

Thanks again for looking over the tutorial! I appreciate it :)

**Comment by TwistedTwigleg** [comment]
Okay, I think I solved the merge conflicts! Hopefully everything is still working and good to go.

**Comment by mhilbrunner** [comment]
Thanks for the huge effort. :) Merging!

File: tutorials/3d/fps_tutorial/part_one.rst [hunk 11]
```diff
@@ -71,213 +86,243 @@ First, go open the project settings and go to the "Input Map" tab. You'll find s
...
         if Input.is_action_pressed("movement_backward"):
-            dir += cam_xform.basis.z.normalized()
+            input_movement_vector.y -= 1
         if Input.is_action_pressed("movement_left"):
-            dir += -cam_xform.basis.x.normalized()
+            input_movement_vector.x -= 1
         if Input.is_action_pressed("movement_right"):
-            dir += cam_xform.basis.x.normalized()
-
+            input_movement_vector.x = 1
```
Comment (Noshyaar) [inline tutorials/3d/fps_tutorial/part_one.rst]: Shouldn't this be `+= 1` ?

Comment (TwistedTwigleg) [reply tutorials/3d/fps_tutorial/part_one.rst]: Yeah, I think so.

Comment (TwistedTwigleg) [reply tutorials/3d/fps_tutorial/part_one.rst]: Ah, okay. Then yeah, that’s fine with me. Thanks!

**Comment by mhilbrunner** [comment]
Cherrypicked into 3.0 as e7e7e3c

**Pull request merged**